runtime/signals_nat.c

```diff
@@ -98,6 +98,37 @@ void caml_garbage_collection(void)
                             nallocs, alloc_len);
 }
 
+void caml_poll(void)
+{
+  frame_descr* d;
```

Comment:
Why does `caml_poll` need to examine frame descriptor at all? Should `caml_poll` directly call one of the functions that processes pending signals? For example, having a quick scan through the code, I would think that `caml_poll` should directly call `caml_raise_if_exception(caml_do_pending_actions_exn ())` as in https://github.com/ocaml/ocaml/blob/trunk/runtime/minor_gc.c#L499.

But that raises an interesting problem. Because poll points are inserted by the compiler even in non-allocating code, exceptions may magically appear out of non-allocating code. Currently, such exceptions can only come from allocations, which can be guarded. My suggestion would be to introduce a global variable (registered as root) to store the exception and in the next allocation point (not poll point), raise that exception. [CC @stedolan]

Reply:
We need the frame descriptor for the spacetime/memprof functions, because they require the pointers to the allocations etc. Just looking at the spacetime/memprof parts to understand this better.

Reply:
The idea is not to call the allocation code at all. In that case, you don't need to examine the frame descriptors. See original comment:

> I would think that caml_poll should directly call caml_raise_if_exception(caml_do_pending_actions_exn ()) as in https://github.com/ocaml/ocaml/blob/trunk/runtime/minor_gc.c#L499.

Reply:
Given the trickiness with asynchronous exceptions, the right thing to do would be to define a function that only does the part of `caml_do_pending_actions_exn` that is known not to raise exceptions https://github.com/ocaml/ocaml/blob/trunk/runtime/signals.c#L293-L302. This should fix the out-of-memory bugs upstream. For multicore, we can add the poll for interrupts in this new function.

Reply:
1. Even I was thinking the poll function can be changed to only perform operations which do not raise exceptions.
2. Given that the poll can trigger a minor collection, I think spacetime/memprof might need to be signalled, regarding the collections? Not very familiar with how memprof/spacetime works, so need to look into this.

Reply:
Reg 2, IINM both spacetime and memprof look at allocations and not collections. If a collection was necessary, then the collection code itself should handle the necessary logic for spacetime and memprof. So it would

Reply:
Regarding exceptions appearing out of nowhere: I don't think exceptions magically appearing in non-allocating code is really worse than exceptions magically appearing in allocations, as happens today. (Both are awful, though, but that's a separate issue!)